File: apps/android/app/src/test/java/video/crumb/app/data/HaControlConfigTest.kt
// SPDX-License-Identifier: AGPL-3.0-or-later

package video.crumb.app.data

import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Per-link control config (migration 0075, issue #440): the Android client must
 * decode `require_confirm` + `allowed_actions` defensively (an older server that
 * omits them decodes exactly as today via the nullable + default fields) and
 * derive the offered action set from `allowed_actions` when present.
 */
class HaControlConfigTest {

    // Mirror the production decoder: tolerant of unknown keys.
    private val json = Json { ignoreUnknownKeys = true }

    @Test
    fun `an older server payload without the fields defaults to today behavior`() {
        val link = json.decodeFromString(
            HaLinkDto.serializer(),
            """{"id":"l1","entity_id":"light.kitchen","role":"actuator","sort_order":0}""",
        )
        assertFalse(link.requireConfirm)
        assertNull(link.allowedActions)
        // Null allowed_actions ⇒ every action permitted; default control set.
        assertTrue(link.actionAllowed("turn_on"))
        // Simple domain collapses to the single primary tap (Toggle), unchanged.
        assertEquals(listOf("toggle"), link.controlActions().map { it.verb })
    }

    @Test
    fun `present fields parse and restrict the offered actions`() {
        val link = json.decodeFromString(
            HaLinkDto.serializer(),
            """{"id":"l2","entity_id":"light.kitchen","role":"actuator","sort_order":0,
                "require_confirm":true,"allowed_actions":["turn_on"]}""",
        )
        assertTrue(link.requireConfirm)
        assertEquals(listOf("turn_on"), link.allowedActions)
        assertTrue(link.actionAllowed("turn_on"))
        assertFalse(link.actionAllowed("turn_off"))
        // Restricted ⇒ present ONLY the permitted verbs (from the full domain set).
        assertEquals(listOf("turn_on"), link.controlActions().map { it.verb })
    }

    @Test
    fun `a light may be restricted to exactly toggle, a default-card superset verb`() {
        val link = json.decodeFromString(
            HaLinkDto.serializer(),
            """{"id":"l3","entity_id":"light.kitchen","role":"actuator","sort_order":0,
                "allowed_actions":["toggle"]}""",
        )
        assertEquals(listOf("toggle"), link.controlActions().map { it.verb })
    }

    @Test
    fun `a cover restricted to open and close drops stop`() {
        val link = json.decodeFromString(
            HaLinkDto.serializer(),
            """{"id":"l4","entity_id":"cover.garage","role":"actuator","sort_order":0,
                "allowed_actions":["open_cover","close_cover"]}""",
        )
        assertEquals(
            listOf("open_cover", "close_cover"),
            link.controlActions().map { it.verb },
        )
    }
}

File: apps/desktop-flutter/test/ha_control_config_test.dart
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// Per-link control config (migration 0075, issue #440): the desktop client must
// parse `require_confirm` + `allowed_actions` defensively (an older server that
// omits them behaves exactly as today) and derive the offered action set from
// `allowed_actions` when present.
//
// Pure, headless assertions on the model + the action-set helper.

import 'package:flutter_test/flutter_test.dart';
import 'package:crumb_desktop/api/ha_models.dart';
import 'package:crumb_desktop/ui/ha_overlay/ha_actions.dart';

void main() {
  group('HaLink control config parsing', () {
    test('an older server payload (no fields) defaults to today behavior', () {
      final l = HaLink.fromJson({
        'id': 'l1',
        'entity_id': 'light.kitchen',
        'role': 'actuator',
        'sort_order': 0,
      });
      expect(l.requireConfirm, isFalse);
      expect(l.allowedActions, isNull);
      // Null allowed_actions ⇒ every action is permitted.
      expect(l.actionAllowed('turn_on'), isTrue);
      expect(l.actionAllowed('turn_off'), isTrue);
    });

    test('present fields are parsed and gate actions', () {
      final l = HaLink.fromJson({
        'id': 'l2',
        'entity_id': 'light.kitchen',
        'role': 'actuator',
        'sort_order': 0,
        'require_confirm': true,
        'allowed_actions': ['turn_on'],
      });
      expect(l.requireConfirm, isTrue);
      expect(l.allowedActions, ['turn_on']);
      expect(l.actionAllowed('turn_on'), isTrue);
      expect(l.actionAllowed('turn_off'), isFalse);
      expect(l.actionAllowed('toggle'), isFalse);
    });
  });

  group('offered action set honors allowed_actions', () {
    // Mirror the layer's _actionsFor logic without the widget: null ⇒ default
    // set, non-null ⇒ full domain set intersected with the permitted verbs.
    List<String> offered(String domain, List<String>? allowed) {
      final actions = allowed == null
          ? haActionsForDomain(domain)
          : haAllActionsForDomain(domain)
              .where((a) => allowed.contains(a.action))
              .toList();
      return actions.map((a) => a.action).toList();
    }

    test('null keeps the default light set (On/Off), unchanged from today', () {
      expect(offered('light', null), ['turn_on', 'turn_off']);
    });

    test('a light restricted to turn_on offers only On', () {
      expect(offered('light', ['turn_on']), ['turn_on']);
    });

    test('a light may be restricted to exactly toggle (superset button)', () {
      expect(offered('light', ['toggle']), ['toggle']);
    });

    test('a cover restricted to open/close drops stop', () {
      expect(
        offered('cover', ['open_cover', 'close_cover']),
        ['open_cover', 'close_cover'],
      );
    });
  });
}
